deep_quoridor/src/quoridor.py

```
@@ -0,0 +1,293 @@
+        self._player_positions = np.array([(0, board_size // 2), (board_size - 1, board_size // 2)])
+        for player, position in enumerate(self._player_positions):
+            self._grid[*position * 2] = player
+        self._walls_remaining = [self.max_walls, self.max_walls]
+
+    def get_player_position(self, player: int) -> Position:
+        """
+        Get the position of the player's pawn.
+        """
+        try:
+            return self._player_positions[player]
+        except IndexError:
+            raise ValueError(f"Invalid player {player}")
+
+    def move_player(self, player: int, new_position: Position):
+        """
+        Move the player's pawn. Doesn't check if the move is valid according to game rules.
+        """
+        new_position = np.asarray(new_position, dtype=np.uint8)
+
+        try:
+            old_position = self._player_positions[player]
+        except IndexError:
+            raise ValueError(f"Invalid player {player}")
+
+        try:
+            self._grid[*old_position * 2] = Board.FREE
+            self._grid[*new_position * 2] = player
+            self._player_positions[player] = new_position
+        except IndexError:
+            return Board.OUT_OF_BOUNDS
+
+    def get_cell(self, position: Position) -> int:
+        """
+        Get the value of a "player cell" in the grid.
+        """
+        position = np.asarray(position, dtype=np.uint8)
+        try:
+            return self._grid[*position * 2]
+        except IndexError:
+            return Board.OUT_OF_BOUNDS
+
+    def add_wall(self, player: int, position: Position, orientation: WallOrientation):
+        """
+        Mark the grid cells corresponding to the wall as occupied.
+        """
+        position = np.asarray(position, dtype=np.uint8)
+
+        if self._walls_remaining[player] < 1:
+            raise RuntimeError("Cannot place wall, no walls remaining for player {player}")
```

Why are we raising when we are out of walls but returning an error code when we are out of bounds? 

Seems like the behavior for failing to place should match regardless of reason.

---

I think in general we could:
- Let exceptions propagate (or raise / assert our own) when something shouldn't happen under normal circumstnaces, e.g. we're trying to acess an invalid position in the grid.  A normal behaving agent shouldn't do it, and for a human player, the position needs to be validated before then.
- Return values when it's the normal flow (e.g. if we're asking if the pawn can be moved somewhere or a wall placed somewhere).
- Raise a custom exception (`CheatException`?) when we're trying to do something that goes against the rules of the game, like moving the pawn to an invalid location or placing a conflicting wall.  The reason for doing that is that this shouldn't be happening anyway, but let's say that the agent has a bug and sends something wrong, we don't want to mask the bug and see an unexpected behavior (e.g. the agent had a turn but nothing happened).

---

I did all of these except the custom exception - I'll leave that to a future PR.